src/store/modules/segmentationModule.js

```diff
@@ -0,0 +1,1411 @@
+
+  const currentImageIdIndex = stackData.currentImageIdIndex;
+  const { rows, columns } = enabledElement.image;
+
+  const numberOfFrames = stackData.imageIds.length;
+  const firstImageId = stackData.imageIds[0];
+
+  let brushStackState = state.series[firstImageId];
+
+  let activeLabelmapIndex;
+
+  if (brushStackState) {
+    activeLabelmapIndex = brushStackState.activeLabelmapIndex;
+
+    if (!brushStackState.labelmaps3D[activeLabelmapIndex]) {
+      _addLabelmap3D(
+        brushStackState,
+        activeLabelmapIndex,
+        rows * columns * numberOfFrames
+      );
+    }
+
+    if (
+      !brushStackState.labelmaps3D[activeLabelmapIndex].labelmaps2D[
+        currentImageIdIndex
+      ]
+    ) {
+      _addLabelmap2DView(
+        brushStackState,
+        activeLabelmapIndex,
+        currentImageIdIndex,
+        rows,
+        columns
+      );
+    }
+  } else {
+    activeLabelmapIndex = 0;
+
+    state.series[firstImageId] = {
+      activeLabelmapIndex,
+      labelmaps3D: [],
+    };
+
+    brushStackState = state.series[firstImageId];
+
+    _addLabelmap3D(
+      brushStackState,
+      activeLabelmapIndex,
+      rows * columns * numberOfFrames
+    );
+
+    _addLabelmap2DView(
+      brushStackState,
+      activeLabelmapIndex,
+      currentImageIdIndex,
+      rows,
+      columns
+    );
+  }
+
+  return {
+    labelmap3D: brushStackState.labelmaps3D[activeLabelmapIndex],
+    currentImageIdIndex,
+    activeLabelmapIndex,
+  };
+}
+
+/**
+ * GetIsSegmentVisible -  Returns if a segment is visible.
+ *
+ * @param  {HTMLElement|string} elementOrEnabledElementUID   The cornerstone enabled
+ *                                                    element or its UUID.
+ * @param  {number} segmentIndex     The segment index.
+ * @param  {number} [labelmapIndex]    If undefined, defaults to the active
+ *                                     labelmap index.
+ * @returns {boolean} True if the segment is visible.
+ */
+function getIsSegmentVisible(
+  elementOrEnabledElementUID,
+  segmentIndex,
+  labelmapIndex
+) {
+  if (!segmentIndex) {
+    return;
+  }
+
+  const element = _getEnabledElement(elementOrEnabledElementUID);
+
+  if (!element) {
+    return;
+  }
+
+  const stackState = getToolState(element, 'stack');
+  const stackData = stackState.data[0];
+  const firstImageId = stackData.imageIds[0];
+
+  const brushStackState = state.series[firstImageId];
+
+  if (!brushStackState) {
+    logger.warn(`brushStackState is undefined`);
+
+    return;
+  }
+
+  if (labelmapIndex === undefined) {
+    labelmapIndex = brushStackState.activeLabelmapIndex;
+  }
+
+  if (!brushStackState.labelmaps3D[labelmapIndex]) {
+    logger.warn(`No labelmap3D of labelmap index ${labelmapIndex} on stack.`);
+
+    return;
+  }
+
+  const labelmap3D = brushStackState.labelmaps3D[labelmapIndex];
+  const visible = labelmap3D.segmentsVisible[segmentIndex];
+
+  return visible || visible === undefined;
```

Comment:
Is there an advantage to having this array instead of an `isVisible` property? If yes, can we default them to visible in the method that creates them so we don't need the undefined check here?

Reply:
A lot of these ugly things spread around here are for the sake of memory management. It stops us from having to initialise `65536` `true`s for each labelmap, which is 4 bytes (GG js...) * 65526 = 256 kilobytes per labelmap.

The user may never switch any off, and may only draw 1 segments for a particular Labelmap.

Maybe this is over-optimisation? But memory is becoming a problem, so avoiding a megabyte of ram for 4 segmentations (for something most users won't use in a session) felt worth it to me?

Reply:
Sometimes the simplest solutions are in plain sight. I'll do this.